### ISSUE TITLE
Allow dict.get(key) check to infer key existence in TypedDict

### ISSUE THREAD

**Issue by kpodp0ra** [microsoft/pyright#9863]
## Describe the solution you’d like

I would like Pyright to recognize that using `if some_dict.get("key"):` is a valid way to check for both the presence and truthiness of an optional key in a `TypedDict`. Currently, Pyright does not infer that this check ensures the key exists before access, leading to a type error when the key is later used.  

While the workaround `if "key" in some_dict and some_dict.get("key"):` works, it is redundant because `dict.get()` already returns `None` if the key is missing. The type checker should recognize that `if some_dict.get("key"):` guarantees safe access, reducing unnecessary checks and improving code clarity.

## Issue

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjALABQDDVANgIYDOHUAEgK4Q2KIgmIAKEWUo0ANLDAw2LALwAxJR2IBKAFwMoBqMCQgOMAPoo2EYjqhmQTRvVLFgUNMQtWbYgBYCQnb8gsIkuvqGSO4BoQB0njBiAETGpt7WxMkR9IZ5UCBefCAoULFCANoA5GlmlplVALoMQA)

```python
from typing import TypedDict


class HumanType(TypedDict, total=False):
    first_name: str


def get_name(human: HumanType):
    if human.get("first_name"):
        return human['first_name']

```

**Could not access item in TypedDict:**  "first_name" is not a required key in "HumanType", so access may result in runtime exception  (reportTypedDictNotRequiredAccess)

## Workaround

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjALABQDDVANgIYDOHUAEgK4Q2KIgmIAKEWUo0ANLDAw2LALwAxJR2IBKAFwMoBqMCQgOMAPoo2EYjqhmQTRvVLFgUNMQtWbYgBYCQnb8gsIkuvqGSO4ARMam3tbEMZgoUAGhUEKk6YEoAHSeMGJxJmaWSTER9Ia1UCBefCBpGUIA2gDk8eU%2BxB0AugxAA)

```python
from typing import TypedDict


class HumanType(TypedDict, total=False):
    first_name: str


def get_name(human: HumanType):
    if "first_name" in human and human.get("first_name"):
        return human['first_name']

```

This works, but it's redudant.

**Comment by kpodp0ra** [comment]
## Related issue

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjADRQDyCMSYKAhgDb3sDOvxPAH14CYgFgAUFKlVOfXlAASAVwjsURMQAotZSjQCUALilRzUYEhC8YQjhGLHGzVh04BtWyAC6M6ZKkxMBQaMR2DsTaABZqGs6q6pokhlAAtAB8UN6mkhaYIbFJAHRhMNoARFY2EeyOFSZm%2BRYg4SogKFBFGh4A5NW29nXEvX55LW0dUBUA4irEthVSQA)

```python
from typing import TypedDict, Optional, assert_type


class HumanType(TypedDict):
    first_name: Optional[str]


def get_name(human: HumanType) -> str:
    if human.get("first_name"):
        return human['first_name']
    return "Guest"

```

**Comment by erictraut** [comment]
Pyright doesn't generally have any hard-coded knowledge about specific classes or methods. It bases its analysis on the type information provided in annotations and on intrinsic behaviors of the language (e.g. the way the `in` operator works). What you're suggesting here would require hard-coding some very specific knowledge about `Mapping.get`.

To be fair, the list of [type guards that pyright supports](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards) does include hard-coded knowledge of a few special functions in the `builtins` namespace: `isinstance`, `issubclass`, and`len`. However, there aren't any situations currently where pyright's type narrowing logic hard-codes knowledge about specific methods in specific classes. I'm reluctant to add such hard-coded logic.

I'll note that other type checkers like mypy likewise don't support the behavior you're suggesting here.

I'm going to reject this enhancement request, but I'm willing to revisit it in the future if there's strong signal from other pyright users (e.g. in the form of thumbs-up reactions in this issue).

In the meantime, I recommend using a temporary variable and a walrus operator if you want to use this coding pattern. It's not only type safe, but it's also more efficient because it doesn't require multiple accesses to the dictionary.

```python
def get_name(human: HumanType):
    if val := human.get("first_name"):
        return val
```

**Comment by fcolecumberri** [comment]
Even better:
```python
human.get("first_name") or "Guest"
```
1 line, no temp variables, compatible with python before `:=` and it almost reads like English.
